jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/Configuration.java

```diff
@@ -39,12 +55,28 @@ public interface Configuration
     void setIdleTimeout(Duration timeout);
 
     /**
-     * Set the Write Timeout.
+     * Set the Frame Write Timeout.
      *
      * @param timeout the timeout duration (timeout &lt;= 0 implies an infinite timeout)
+     * @deprecated use {@link #setFrameWriteTimeout(Duration)} or {@link #setMessageWriteTimeout(Duration)}.
      */
+    @Deprecated(since = "12.0.21", forRemoval = true)
     void setWriteTimeout(Duration timeout);
 
+    /**
+     * Set the Frame Write Timeout.
+     *
+     * @param timeout the timeout duration (timeout &lt;= 0 implies an infinite timeout).
+     */
+    void setFrameWriteTimeout(Duration timeout);
```

Comment:
I don't think this is the right approach, as it is the same model as the Jakarta APIs, and it's subject to concurrency issues, etc.

I would store the `sendTimeout` in the Jakarta implementation (e.g. `org.eclipse.jetty.ee11.websocket.jakarta.common.JakartaWebSocketAsyncRemote`), and then call `coreSession.sendFrame(frame, callback, batch, timeout)`.

In this way, the WebSocket core APIs will remain "clean" WRT to concurrency.

Reply:
This is a very large change, I think it should be considered as a separate PR / Issue.

The Jakarta API doesn't use a Frame write timeout, it does a message timeout.

So we would probably want to change `sendFrame(frame, callback, batch)` to be `sendFrame(FrameEntry)`, which has the different timeout configurations. This change in signature would need to be applied to several flushers and the extension stack.

We also have to maintain the default timeout values for the connection. These can be modified through setters on the `CoreSession` or `WebSocketNegotiator`. And work as the default timeout values for any frame/message sent on the connection.